pilot/pkg/serviceregistry/external/servicediscovery.go

```diff
@@ -260,6 +282,65 @@ func (d *ServiceEntryStore) InstancesByPort(svc *model.Service, port int,
 	return out, nil
 }
 
+// servicesWithEntry contains a ServiceEntry and associated model.Services
+// This is used only as a key to a map, not intended for external usage
+type servicesWithEntry struct {
+	entry    *networking.ServiceEntry
+	services []*model.Service
+}
+
+// edsUpdate triggers an EDS update for the given instances
+func (d *ServiceEntryStore) edsUpdate(instances []*model.ServiceInstance) {
+	allInstances := []*model.ServiceInstance{}
+
+	// Find all keys we need to lookup
+	keys := map[instancesKey]struct{}{}
+	for _, i := range instances {
+		keys[makeInstanceKey(i)] = struct{}{}
+	}
+
+	d.maybeRefreshIndexes()
+
+	d.storeMutex.RLock()
+	for key := range keys {
+		for _, i := range d.instances[key] {
+			allInstances = append(allInstances, i...)
+		}
+	}
+	d.storeMutex.RUnlock()
+
+	// This was a delete
+	if len(allInstances) == 0 {
+		for k := range keys {
+			_ = d.XdsUpdater.EDSUpdate(d.Cluster(), string(k.hostname), k.namespace, nil)
+		}
+		return
+	}
+
+	endpoints := make(map[instancesKey][]*model.IstioEndpoint)
+	for _, instance := range allInstances {
+		port := instance.ServicePort
+		key := makeInstanceKey(instance)
+		endpoints[key] = append(endpoints[key],
+			&model.IstioEndpoint{
+				Address:         instance.Endpoint.Address,
+				EndpointPort:    uint32(port.Port),
+				ServicePortName: port.Name,
+				Labels:          instance.Endpoint.Labels,
+				UID:             instance.Endpoint.UID,
+				ServiceAccount:  instance.Endpoint.ServiceAccount,
+				Network:         instance.Endpoint.Network,
+				Locality:        instance.Endpoint.Locality,
+				LbWeight:        instance.Endpoint.LbWeight,
+				TLSMode:         instance.Endpoint.TLSMode,
+			})
+	}
```

Comment:
nit: we should prevent such use, it is very likely to miss a field. Abstract it as a function and reuse as much as possible